src/vfs2.c

```diff
@@ -1311,23 +1393,37 @@ static int open_entry(struct common *common, const char *name, struct entry *e)
 	rv = link(e->wal_cur_fixed_name, e->wal_moving_name);
 	(void)rv;
 
-	rv = shm_init(&e->shm);
-	*get_full_hdr(e) = initial_full_hdr(hdr_cur);
-
-	e->wal_cursor = wal_cursor_from_size(e->page_size, size_cur);
-
-	int next = WTX_EMPTY;
-	if (size_cur >=
-	    wal_offset_from_cursor(0 /* this doesn't matter */, 0)) {
-		/* TODO verify the header here */
+	/* If WAL-cur contains a valid header, walk it to initialize wal_cursor
+	 * and the WAL-index. Also, set the page size.
+	 *
+	 * If WAL-cur is empty, then we are starting up for the first time, and
+	 * we don't initialize the WAL-index. It will be initialized either by
+	 * SQLite running recovery or by add_uncommitted, whichever happens
+	 * first. (In general, we want to avoid letting SQLite run recovery, but
+	 * in this case it's harmless, since if the WAL is empty then it can't
+	 * read any uncommitted data.)
+	 */
+	if (size_cur > 0) {
 		e->page_size = ByteGetBe32(hdr_cur.page_size);
-		next = WTX_BASE;
+		rv = shm_init(&e->shm, hdr_cur);
+		if (rv != SQLITE_OK) {
+			return rv;
+		}
+		rv = walk_wal(e->wal_cur, size_cur, hdr_cur, NULL,
+			      &e->wal_cursor, &e->shm);
+		if (rv != SQLITE_OK) {
+			return rv;
+		}
 	}
-	if (size_cur >= wal_offset_from_cursor(e->page_size, 1)) {
+	/* If we found at least one valid transaction in the WAL, take the write
+	 * lock. */
+	if (e->wal_cursor > 0) {
 		e->shm_locks[WAL_WRITE_LOCK] = VFS2_EXCLUSIVE;
-		next = WTX_FOLLOWING;
 	}
-	sm_move(&e->wtx_sm, next);
+
+	sm_move(&e->wtx_sm, e->wal_cursor > 0 ? WTX_FOLLOWING
+			    : size_cur > 0    ? WTX_BASE
+					      : WTX_EMPTY);
```

Comment:
At startup we consider all the transactions in WAL-cur to be uncommitted, so if `walk_wal` finds any transactions we need to be in FOLLOWING state. Since FLUSH was removed both leaders and followers can be in BASE---I need to update the comment on the enum.